provider.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2020 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package blueprint
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// This file implements Providers, modelled after Bazel
+// (https://docs.bazel.build/versions/master/skylark/rules.html#providers).
+// Each provider can be associated with a mutator, in which case the value for the provider for a
```

Comment:
What's the use case you're trying to support (or anti-patterns you're trying to disallow) by coupling providers with a mutator or gba? Should not specifying and thus allowing a provider to be set in any phase be an option?

Reply:
Oops, accidentally marked this resolved.

I was worried about ordering.  Since it is valid for a provider to exist on some modules and not on others, and it is often very hard to follow the mutator ordering in Soong, I thought it would be common to read a provider from a module before the mutator that was supposed to set it, get the zero value, and be confused why it was always empty/unset.  This way the logic on whether or not a provider is set is guaranteed to be entirely contained within a single mutator.